src/wagtail_personalisation/rules.py

```diff
@@ -220,18 +226,37 @@ class VisitCountRule(AbstractBaseRule):
     class Meta:
         verbose_name = _('Visit count Rule')
 
+    def _get_user_session(self, user):
+        sessions = Session.objects.iterator()
+        for session in sessions:
+            session_data = session.get_decoded()
+            if session_data.get('_auth_user_id') == str(user.id):
+                return SessionStore(session_key=session.session_key)
+        return SessionStore()
+
     def test_user(self, request, user=None):
+        # Local import for cyclic import
+        from wagtail_personalisation.adapters import (
+            get_segment_adapter, SessionSegmentsAdapter, SEGMENT_ADAPTER_CLASS)
+
         if user:
-            # This rule currently does not support testing a user directly
-            # TODO: Make this test a user directly when the rule uses
-            # historical data
+            # Create a fake request so we can use the adapter
+            request = RequestFactory().get('/')
+            request.user = user
+
+            # If we're using the session adapter check for an active session
+            if SEGMENT_ADAPTER_CLASS == SessionSegmentsAdapter:
+                request.session = self._get_user_session(user)
```

Comment:
I just wanted to check quickly what the impact of this will be for the `SessionSegmentsAdapter` even though I don't think we're using it anywhere right now.

In Indonesia we have 78902 Session objects and it takes between 8 and 11 seconds to iterate over them and decode each one, which I think is what will happen in the worst case.

As the `test_user()` method for each rule is called on every page load I think this could add 8 seconds to a page load for anything using the `VisitCountRule`.

I don't feel too strongly either way as the code is unused but I'm tempted to suggest we shouldn't merge in this code if we're not going to use it and it would perform poorly in production.

Reply:
@alexmuller when a page is loaded the `test_user()` method is called with a request, not with a user. Since this code is only run when we call it with a user it won't get executed. This will currently get run when an admin downloads the users for a segment or when check how many users match a segment they are busy creating.

I can change it to only run if there isn't a request rather than running whenever there's a user if you'd prefer?

Reply:
Ah yeah you're right, it would only run in that case. I'm not sure what the timeouts are on the loadbalancer but I imagine it's not going to significantly longer than 10 seconds so there might be an issue there. But in any case let's give it a go and see what happens.